File: tests/generate.test.ts
import { describe, expect, it } from 'vitest';
import { addVec3, ORIENTATION_COUNT, vec3, vec3Key, type Vec3 } from '../src/core/grid';
import { placedVoxels, type Piece, type Placement } from '../src/core/piece';
import {
  generatePieces,
  generatePuzzle,
  MAX_SPACE_SIZE,
  maxPieces,
  MIN_PIECE_COUNT,
  MIN_SPACE_SIZE,
  scatterPlacements,
  solutionPlacements,
} from '../src/core/generate';
import { isSolved } from '../src/core/solve';

/** テスト側で独立に定義する 6 近傍。 */
const NEIGHBORS: readonly Vec3[] = [
  vec3(1, 0, 0),
  vec3(-1, 0, 0),
  vec3(0, 1, 0),
  vec3(0, -1, 0),
  vec3(0, 0, 1),
  vec3(0, 0, -1),
];

const SIZES: readonly number[] = [3, 4, 5, 6, 7];
const SEEDS: readonly number[] = [1, 7, 12345, 2026];

/** M の代表値（下限・中間・上限）。 */
function representativePieceCounts(n: number): number[] {
  const limit = maxPieces(n);
  const middle = Math.floor((MIN_PIECE_COUNT + limit) / 2);
  return [...new Set([MIN_PIECE_COUNT, middle, limit])];
}

/** ボクセル集合が 6 近傍で連結しているか（BFS）。 */
function isConnected(voxels: readonly Vec3[]): boolean {
  const start = voxels[0];
  if (start === undefined) return false;
  const all = new Set(voxels.map(vec3Key));
  const seen = new Set<string>([vec3Key(start)]);
  const queue: Vec3[] = [start];
  for (let head = 0; head < queue.length; head++) {
    const v = queue[head];
    if (v === undefined) continue;
    for (const offset of NEIGHBORS) {
      const neighbor = addVec3(v, offset);
      const key = vec3Key(neighbor);
      if (!all.has(key) || seen.has(key)) continue;
      seen.add(key);
      queue.push(neighbor);
    }
  }
  return seen.size === all.size;
}

/** ピースと配置を id で対応付けて、全ワールド座標を集める。 */
function worldVoxels(pieces: readonly Piece[], placements: readonly Placement[]): Vec3[] {
  const world: Vec3[] = [];
  for (const placement of placements) {
    const piece = pieces.find((p): boolean => p.id === placement.pieceId);
    if (piece === undefined) throw new Error(`テスト: ピース ${placement.pieceId} が無い`);
    for (const v of placedVoxels(piece, placement)) world.push(v);
  }
  return world;
}

describe('maxPieces', () => {
  it('SPEC.md 3.1 の目安（N=3 で 6、N=7 で 40）に一致する', () => {
    expect(maxPieces(3)).toBe(6);
    expect(maxPieces(4)).toBe(16);
    expect(maxPieces(5)).toBe(31);
    expect(maxPieces(6)).toBe(40);
    expect(maxPieces(7)).toBe(40);
  });

  it('どの N でも下限 2 以上で N³ 以下', () => {
    for (const n of SIZES) {
      expect(maxPieces(n)).toBeGreaterThanOrEqual(MIN_PIECE_COUNT);
      expect(maxPieces(n)).toBeLessThanOrEqual(n ** 3);
    }
  });

  it('範囲外の N は RangeError', () => {
    for (const n of [MIN_SPACE_SIZE - 1, MAX_SPACE_SIZE + 1, 0, -3, 3.5, NaN, Infinity]) {
      expect(() => maxPieces(n)).toThrow(RangeError);
    }
  });
});

describe('generatePuzzle', () => {
  for (const n of SIZES) {
    for (const m of representativePieceCounts(n)) {
      for (const seed of SEEDS) {
        it(`N=${n} / M=${m} / seed=${seed}: 全ボクセルがちょうど 1 つのピースに属し、各ピースが連結`, () => {
          const puzzle = generatePuzzle(n, m, seed);
          expect(puzzle.n).toBe(n);
          expect(puzzle.m).toBe(m);
          expect(puzzle.seed).toBe(seed);
          expect(puzzle.pieces).toHaveLength(m);
          expect(puzzle.solution).toHaveLength(m);

          // ピース id は 0..M-1
          expect(puzzle.pieces.map((p): number => p.id)).toEqual(
            Array.from({ length: m }, (_, i): number => i),
          );

          const occurrence = new Map<string, number>();
          for (let i = 0; i < m; i++) {
            const piece = puzzle.pieces[i];
            const placement = puzzle.solution[i];
            if (piece === undefined || placement === undefined) throw new Error('ピースが無い');

            // 局所原点が (0,0,0) に正規化されている
            expect(piece.voxels.some((v): boolean => v.x === 0 && v.y === 0 && v.z === 0)).toBe(
              true,
            );
            // 解答配置は向きが恒等
            expect(placement.orientation).toBe(0);

            const world = placedVoxels(piece, placement);
            expect(world.length).toBeGreaterThan(0);
            // 6 近傍で連結している
            expect(isConnected(world)).toBe(true);
            // 解答位置は N×N×N の中に収まる
            const inside = world.every(
              (v): boolean =>
                v.x >= 0 && v.x < n && v.y >= 0 && v.y < n && v.z >= 0 && v.z < n,
            );
            expect(inside).toBe(true);

            for (const v of world) {
              const key = vec3Key(v);
              occurrence.set(key, (occurrence.get(key) ?? 0) + 1);
            }
          }

          // 欠けなし（N³ マスすべて）・重複なし（どのマスもちょうど 1 回）
          expect(occurrence.size).toBe(n ** 3);
          expect([...occurrence.values()].every((c): boolean => c === 1)).toBe(true);
        });
      }
    }
  }

  it('同じ引数なら結果が完全に一致する（決定的）', () => {
    for (const n of SIZES) {
      for (const m of representativePieceCounts(n)) {
        expect(generatePuzzle(n, m, 4649)).toEqual(generatePuzzle(n, m, 4649));
      }
    }
  });

  it('シードが違えば別の分割になる', () => {
    const a = generatePuzzle(5, 8, 1);
    const b = generatePuzzle(5, 8, 2);
    expect(a).not.toEqual(b);
  });

  it('不正な N は RangeError', () => {
    for (const n of [MIN_SPACE_SIZE - 1, MAX_SPACE_SIZE + 1, 0, -1, 4.5, NaN]) {
      expect(() => generatePuzzle(n, 2, 1)).toThrow(RangeError);
    }
  });

  it('不正な M は RangeError', () => {
    for (const n of SIZES) {
      expect(() => generatePuzzle(n, MIN_PIECE_COUNT - 1, 1)).toThrow(RangeError);
      expect(() => generatePuzzle(n, 0, 1)).toThrow(RangeError);
      expect(() => generatePuzzle(n, -2, 1)).toThrow(RangeError);
      expect(() => generatePuzzle(n, maxPieces(n) + 1, 1)).toThrow(RangeError);
      expect(() => generatePuzzle(n, 2.5, 1)).toThrow(RangeError);
      expect(() => generatePuzzle(n, NaN, 1)).toThrow(RangeError);
    }
  });

  it('不正なシードは RangeError', () => {
    expect(() => generatePuzzle(3, 2, 1.5)).toThrow(RangeError);
    expect(() => generatePuzzle(3, 2, NaN)).toThrow(RangeError);
  });
});

describe('generatePieces / solutionPlacements', () => {
  it('generatePuzzle と同じものを返す', () => {
    for (const n of SIZES) {
      const m = representativePieceCounts(n)[1] ?? MIN_PIECE_COUNT;
      const puzzle = generatePuzzle(n, m, 31);
      expect(generatePieces(n, m, 31)).toEqual(puzzle.pieces);
      expect(solutionPlacements(n, m, 31)).toEqual(puzzle.solution);
    }
  });

  it('返した配列を書き換えても次の呼び出しに影響しない', () => {
    const pieces = generatePieces(3, 4, 5);
    pieces.pop();
    expect(generatePieces(3, 4, 5)).toHaveLength(4);
  });
});

describe('scatterPlacements', () => {
  for (const n of SIZES) {
    for (const m of [MIN_PIECE_COUNT, maxPieces(n)]) {
      for (const seed of [1, 99]) {
        it(`N=${n} / M=${m} / seed=${seed}: 重ならず、向きは 0..23、クリアではない`, () => {
          const pieces = generatePieces(n, m, seed);
          const placements = scatterPlacements(pieces, n, seed + 1);

          expect(placements).toHaveLength(m);
          expect([...placements].map((p): number => p.pieceId).sort((a, b): number => a - b)).toEqual(
            pieces.map((p): number => p.id),
          );

          const validOrientation = placements.every(
            (p): boolean =>
              Number.isInteger(p.orientation) &&
              p.orientation >= 0 &&
              p.orientation < ORIENTATION_COUNT,
          );
          expect(validOrientation).toBe(true);

          const validPosition = placements.every(
            (p): boolean =>
              Number.isInteger(p.position.x) &&
              Number.isInteger(p.position.y) &&
              Number.isInteger(p.position.z),
          );
          expect(validPosition).toBe(true);

          // ピース同士が重ならない（ワールド座標の重複が無い）
          const world = worldVoxels(pieces, placements);
          expect(world).toHaveLength(n ** 3);
          expect(new Set(world.map(vec3Key)).size).toBe(n ** 3);

          // 散らした直後にクリアになっていない
          expect(isSolved(pieces, placements, n)).toBe(false);
        });
      }
    }
  }

  it('同じ seed なら同じ配置（決定的）', () => {
    for (const n of SIZES) {
      const pieces = generatePieces(n, maxPieces(n), 3);
      expect(scatterPlacements(pieces, n, 777)).toEqual(scatterPlacements(pieces, n, 777));
    }
  });

  it('seed が違えば別の配置になる', () => {
    const pieces = generatePieces(4, 6, 3);
    expect(scatterPlacements(pieces, 4, 1)).not.toEqual(scatterPlacements(pieces, 4, 2));
  });

  it('散らす範囲は立方体の周囲 ±(N+2) 程度に収まる', () => {
    for (const n of SIZES) {
      const pieces = generatePieces(n, maxPieces(n), 11);
      const placements = scatterPlacements(pieces, n, 12);
      const center = Math.floor((n - 1) / 2);
      const half = n + 2;
      const inRange = placements.every(
        (p): boolean =>
          Math.abs(p.position.x - center) <= half &&
          Math.abs(p.position.y - center) <= half &&
          Math.abs(p.position.z - center) <= half,
      );
      expect(inRange).toBe(true);
    }
  });

  it('不正な引数は RangeError', () => {
    const pieces = generatePieces(3, 2, 1);
    expect(() => scatterPlacements(pieces, 2, 1)).toThrow(RangeError);
    expect(() => scatterPlacements(pieces, 8, 1)).toThrow(RangeError);
    expect(() => scatterPlacements(pieces, 3, 0.5)).toThrow(RangeError);
    expect(() => scatterPlacements([], 3, 1)).toThrow(RangeError);
  });
});

File: Source/CUBELITHCore/Private/Tests/GenerateTest.cpp
// WebMock/tests/generate.test.ts の maxPieces / generatePuzzle / generatePieces・solutionPlacements の正常系の移植
// 移植しないテスト: maxPieces の「範囲外の N は RangeError」（checkf で停止するため）
// 移植しないテスト: generatePuzzle の「不正な N は RangeError」（checkf で停止するため）
// 移植しないテスト: generatePuzzle の「不正な M は RangeError」（checkf で停止するため）
// 移植しないテスト: generatePuzzle の「不正なシードは RangeError」（checkf で停止するため）
// 「不正なシード（非整数）」は C++ では型で排除される（シードを uint32 で受けるので整数しか渡せない）
// 006 で足すテスト: scatterPlacements と scatterPlacements のオプション（ScatterPlacements が要る）
// 組み合わせを回すテストは、一致しているときは記録を増やさず、食い違ったときだけ AddError する

#include "Misc/AutomationTest.h"

#if WITH_DEV_AUTOMATION_TESTS

#include "Generate.h"
#include "Grid.h"
#include "Piece.h"

namespace CubelithCoreTests
{
	// GenerateTest.cpp 専用のヘルパ。unity ビルドでは他のテストファイルと同じ翻訳単位に入るので、名前をこの中に閉じる
	namespace GenerateTestDetail
	{
		/** 試す空間サイズ（TS の SIZES） */
		const TArray<int32>& Sizes()
		{
			static const TArray<int32> Values{ 3, 4, 5, 6, 7 };
			return Values;
		}

		/** 試すシード（TS の SEEDS） */
		const TArray<uint32>& Seeds()
		{
			static const TArray<uint32> Values{ 1, 7, 12345, 2026 };
			return Values;
		}

		/** M の代表値（下限・中間・上限。TS の representativePieceCounts）。重複は取り除く */
		TArray<int32> RepresentativePieceCounts(int32 N)
		{
			const int32 Limit = Cubelith::MaxPieces(N);
			const int32 Middle = (Cubelith::MinPieceCount + Limit) / 2;

			TArray<int32> Result;
			for (const int32 Value : { Cubelith::MinPieceCount, Middle, Limit })
			{
				Result.AddUnique(Value);
			}
			return Result;
		}

		/** ボクセル集合が 6 近傍で連結しているか（BFS。TS の isConnected） */
		bool IsConnected(const TArray<Cubelith::FVec3>& Voxels)
		{
			if (Voxels.Num() == 0)
			{
				return false;
			}

			static const TArray<Cubelith::FVec3> Neighbors{
				Cubelith::Vec3(1, 0, 0), Cubelith::Vec3(-1, 0, 0), Cubelith::Vec3(0, 1, 0),
				Cubelith::Vec3(0, -1, 0), Cubelith::Vec3(0, 0, 1), Cubelith::Vec3(0, 0, -1) };

			const TSet<Cubelith::FVec3> All(Voxels);
			TSet<Cubelith::FVec3> Seen;
			Seen.Add(Voxels[0]);
			TArray<Cubelith::FVec3> Queue{ Voxels[0] };
			for (int32 Head = 0; Head < Queue.Num(); ++Head)
			{
				const Cubelith::FVec3 V = Queue[Head];
				for (const Cubelith::FVec3& Offset : Neighbors)
				{
					const Cubelith::FVec3 Neighbor = Cubelith::AddVec3(V, Offset);
					if (!All.Contains(Neighbor) || Seen.Contains(Neighbor))
					{
						continue;
					}
					Seen.Add(Neighbor);
					Queue.Add(Neighbor);
				}
			}
			return Seen.Num() == All.Num();
		}

		/** ピースが一致するか（id とボクセルの並び） */
		bool PieceEquals(const Cubelith::FPiece& A, const Cubelith::FPiece& B)
		{
			return A.Id == B.Id && A.Voxels == B.Voxels;
		}

		/** ピースの配列が一致するか */
		bool PiecesEqual(const TArray<Cubelith::FPiece>& A, const TArray<Cubelith::FPiece>& B)
		{
			if (A.Num() != B.Num())
			{
				return false;
			}
			for (int32 Index = 0; Index < A.Num(); ++Index)
			{
				if (!PieceEquals(A[Index], B[Index]))
				{
					return false;
				}
			}
			return true;
		}

		/** 生成結果が完全に一致するか（TS の toEqual に当たる比較） */
		bool PuzzleEquals(const Cubelith::FGeneratedPuzzle& A, const Cubelith::FGeneratedPuzzle& B)
		{
			return A.N == B.N && A.M == B.M && A.Seed == B.Seed &&
				PiecesEqual(A.Pieces, B.Pieces) && A.Solution == B.Solution;
		}

		/** N / M / シードの組み合わせをメッセージに出す */
		FString Describe(int32 N, int32 M, uint32 Seed)
		{
			return FString::Printf(TEXT("N=%d / M=%d / seed=%u"), N, M, Seed);
		}

		/**
		 * 1 つの組み合わせについて「全ボクセルがちょうど 1 つのピースに属し、各ピースが連結」を確かめる。
		 * 食い違ったら AddError して false（その組み合わせの残りは見ない）
		 */
		bool CheckPartition(FAutomationTestBase& Test, int32 N, int32 M, uint32 Seed)
		{
			const FString What = Describe(N, M, Seed);
			const Cubelith::FGeneratedPuzzle Puzzle = Cubelith::GeneratePuzzle(N, M, Seed);

			if (Puzzle.N != N || Puzzle.M != M || Puzzle.Seed != Seed)
			{
				Test.AddError(FString::Printf(TEXT("%s: 生成結果の N / M / シードが引数と違う（%d / %d / %u）"),
					*What, Puzzle.N, Puzzle.M, Puzzle.Seed));
				return false;
			}
			if (Puzzle.Pieces.Num() != M || Puzzle.Solution.Num() != M)
			{
				Test.AddError(FString::Printf(TEXT("%s: 件数が M でない（pieces %d / solution %d）"),
					*What, Puzzle.Pieces.Num(), Puzzle.Solution.Num()));
				return false;
			}

			TMap<Cubelith::FVec3, int32> Occurrence;
			Occurrence.Reserve(N * N * N);
			for (int32 Index = 0; Index < M; ++Index)
			{
				const Cubelith::FPiece& Piece = Puzzle.Pieces[Index];
				const Cubelith::FPlacement& Placement = Puzzle.Solution[Index];

				// ピース id は 0..M-1
				if (Piece.Id != Index)
				{
					Test.AddError(FString::Printf(TEXT("%s: pieces[%d].id が %d（期待 %d）"), *What, Index, Piece.Id, Index));
					return false;
				}
				// 局所原点が (0,0,0) に正規化されている
				if (!Piece.Voxels.Contains(Cubelith::Vec3(0, 0, 0)))
				{
					Test.AddError(FString::Printf(TEXT("%s: pieces[%d] に局所原点 (0,0,0) が無い"), *What, Index));
					return false;
				}
				// 解答配置は向きが恒等
				if (Placement.Orientation != Cubelith::IdentityOrientation)
				{
					Test.AddError(FString::Printf(TEXT("%s: solution[%d] の向きが %d（期待 0）"), *What, Index, Placement.Orientation));
					return false;
				}

				const TArray<Cubelith::FVec3> World = Cubelith::PlacedVoxels(Piece, Placement);
				if (World.Num() == 0)
				{
					Test.AddError(FString::Printf(TEXT("%s: pieces[%d] のボクセルが空"), *What, Index));
					return false;
				}
				// 6 近傍で連結している
				if (!IsConnected(World))
				{
					Test.AddError(FString::Printf(TEXT("%s: pieces[%d] が連結していない"), *What, Index));
					return false;
				}

				for (const Cubelith::FVec3& V : World)
				{
					// 解答位置は N×N×N の中に収まる
					if (V.X < 0 || V.X >= N || V.Y < 0 || V.Y >= N || V.Z < 0 || V.Z >= N)
					{
						Test.AddError(FString::Printf(TEXT("%s: pieces[%d] の解答位置 (%d,%d,%d) が立方体の外"),
							*What, Index, V.X, V.Y, V.Z));
						return false;
					}
					++Occurrence.FindOrAdd(V, 0);
				}
			}

			// 欠けなし（N³ マスすべて）・重複なし（どのマスもちょうど 1 回）
			const int32 Total = N * N * N;
			if (Occurrence.Num() != Total)
			{
				Test.AddError(FString::Printf(TEXT("%s: 埋まったマスが %d 個（期待 %d 個）"), *What, Occurrence.Num(), Total));
				return false;
			}
			for (const TPair<Cubelith::FVec3, int32>& Pair : Occurrence)
			{
				if (Pair.Value != 1)
				{
					Test.AddError(FString::Printf(TEXT("%s: マス (%d,%d,%d) が %d 個のピースに属している"),
						*What, Pair.Key.X, Pair.Key.Y, Pair.Key.Z, Pair.Value));
					return false;
				}
			}

			return true;
		}
	}
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FCubelithGenerateMaxPiecesMatchesRuleTest, "CUBELITH.Core.Generate.MaxPiecesMatchesRule",
	EAutomationTestFlags_ApplicationContextMask | EAutomationTestFlags::ProductFilter)

bool FCubelithGenerateMaxPiecesMatchesRuleTest::RunTest(const FString& Parameters)
{
	// RULES.md 3.1 の上限 N + 4(N−2) に一致する
	TestEqual(TEXT("MaxPieces(3)"), Cubelith::MaxPieces(3), 7);
	TestEqual(TEXT("MaxPieces(4)"), Cubelith::MaxPieces(4), 12);
	TestEqual(TEXT("MaxPieces(5)"), Cubelith::MaxPieces(5), 17);
	TestEqual(TEXT("MaxPieces(6)"), Cubelith::MaxPieces(6), 22);
	TestEqual(TEXT("MaxPieces(7)"), Cubelith::MaxPieces(7), 27);

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FCubelithGenerateMaxPiecesWithinBoundsTest, "CUBELITH.Core.Generate.MaxPiecesWithinBounds",
	EAutomationTestFlags_ApplicationContextMask | EAutomationTestFlags::ProductFilter)

bool FCubelithGenerateMaxPiecesWithinBoundsTest::RunTest(const FString& Parameters)
{
	using namespace CubelithCoreTests;

	// どの N でも下限 2 以上で N³ 以下
	for (const int32 N : GenerateTestDetail::Sizes())
	{
		const int32 Limit = Cubelith::MaxPieces(N);
		TestTrue(FString::Printf(TEXT("MaxPieces(%d)=%d が下限 %d 以上"), N, Limit, Cubelith::MinPieceCount),
			Limit >= Cubelith::MinPieceCount);
		TestTrue(FString::Printf(TEXT("MaxPieces(%d)=%d が N³ 以下"), N, Limit), Limit <= N * N * N);
	}

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FCubelithGeneratePartitionsCubeTest, "CUBELITH.Core.Generate.PartitionsCube",
	EAutomationTestFlags_ApplicationContextMask | EAutomationTestFlags::ProductFilter)

bool FCubelithGeneratePartitionsCubeTest::RunTest(const FString& Parameters)
{
	using namespace CubelithCoreTests;

	// N × M の代表値 × シードの全組み合わせで「全ボクセルがちょうど 1 つのピースに属し、各ピースが連結」
	int32 Checked = 0;
	for (const int32 N : GenerateTestDetail::Sizes())
	{
		for (const int32 M : GenerateTestDetail::RepresentativePieceCounts(N))
		{
			for (const uint32 Seed : GenerateTestDetail::Seeds())
			{
				GenerateTestDetail::CheckPartition(*this, N, M, Seed);
				++Checked;
			}
		}
	}

	// 5 サイズ × M の代表値 3 通り × シード 4 個
	TestEqual(TEXT("確かめた組み合わせの数"), Checked, 60);

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FCubelithGenerateDeterministicTest, "CUBELITH.Core.Generate.Deterministic",
	EAutomationTestFlags_ApplicationContextMask | EAutomationTestFlags::ProductFilter)

bool FCubelithGenerateDeterministicTest::RunTest(const FString& Parameters)
{
	using namespace CubelithCoreTests;

	// 同じ引数なら結果が完全に一致する（決定的）
	for (const int32 N : GenerateTestDetail::Sizes())
	{
		for (const int32 M : GenerateTestDetail::RepresentativePieceCounts(N))
		{
			const bool bEquals = GenerateTestDetail::PuzzleEquals(
				Cubelith::GeneratePuzzle(N, M, 4649), Cubelith::GeneratePuzzle(N, M, 4649));
			if (!bEquals)
			{
				AddError(FString::Printf(TEXT("%s: 2 回の生成結果が違う"), *GenerateTestDetail::Describe(N, M, 4649)));
			}
		}
	}

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FCubelithGenerateDifferentSeedDiffersTest, "CUBELITH.Core.Generate.DifferentSeedDiffers",
	EAutomationTestFlags_ApplicationContextMask | EAutomationTestFlags::ProductFilter)

bool FCubelithGenerateDifferentSeedDiffersTest::RunTest(const FString& Parameters)
{
	using namespace CubelithCoreTests;

	// シードが違えば別の分割になる
	TestFalse(TEXT("seed=1 と seed=2 の生成結果が違う"),
		GenerateTestDetail::PuzzleEquals(Cubelith::GeneratePuzzle(5, 8, 1), Cubelith::GeneratePuzzle(5, 8, 2)));

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FCubelithGeneratePiecesMatchPuzzleTest, "CUBELITH.Core.Generate.PiecesMatchPuzzle",
	EAutomationTestFlags_ApplicationContextMask | EAutomationTestFlags::ProductFilter)

bool FCubelithGeneratePiecesMatchPuzzleTest::RunTest(const FString& Parameters)
{
	using namespace CubelithCoreTests;

	// GeneratePieces / SolutionPlacements は GeneratePuzzle と同じものを返す
	for (const int32 N : GenerateTestDetail::Sizes())
	{
		const TArray<int32> Counts = GenerateTestDetail::RepresentativePieceCounts(N);
		const int32 M = Counts.IsValidIndex(1) ? Counts[1] : Cubelith::MinPieceCount;
		const Cubelith::FGeneratedPuzzle Puzzle = Cubelith::GeneratePuzzle(N, M, 31);

		TestTrue(FString::Printf(TEXT("%s: GeneratePieces が GeneratePuzzle と同じ"), *GenerateTestDetail::Describe(N, M, 31)),
			GenerateTestDetail::PiecesEqual(Cubelith::GeneratePieces(N, M, 31), Puzzle.Pieces));
		TestTrue(FString::Printf(TEXT("%s: SolutionPlacements が GeneratePuzzle と同じ"), *GenerateTestDetail::Describe(N, M, 31)),
			Cubelith::SolutionPlacements(N, M, 31) == Puzzle.Solution);
	}

	return true;
}

IMPLEMENT_SIMPLE_AUTOMATION_TEST(FCubelithGenerateReturnsCopyTest, "CUBELITH.Core.Generate.ReturnsCopy",
	EAutomationTestFlags_ApplicationContextMask | EAutomationTestFlags::ProductFilter)

bool FCubelithGenerateReturnsCopyTest::RunTest(const FString& Parameters)
{
	// 返した配列を書き換えても次の呼び出しに影響しない
	TArray<Cubelith::FPiece> Pieces = Cubelith::GeneratePieces(3, 4, 5);
	TestEqual(TEXT("最初の呼び出しの件数"), Pieces.Num(), 4);
	Pieces.RemoveAt(Pieces.Num() - 1);

	TestEqual(TEXT("書き換えた後の呼び出しの件数"), Cubelith::GeneratePieces(3, 4, 5).Num(), 4);

	return true;
}

#endif // WITH_DEV_AUTOMATION_TESTS
